ffi/src/lib.rs

```diff
@@ -81,8 +76,13 @@ pub trait IntoFfi: Sized {
     /// The resulting type after conversion
     type Target: ReprC;
 
+    /// Type into which state can be stored during conversion. Useful for returning
+    /// non-owning types but performing some conversion which requires allocation.
+    /// Serves similar purpose as does context in a closure
+    type Store;
+
     /// Convert from [`Self`] into [`Self::Target`]
-    fn into_ffi(self) -> Self::Target;
+    fn into_ffi(self, store: &mut core::mem::MaybeUninit<Self::Store>) -> Self::Target;
```

Comment:
At first I thought that I'll have to use `MaybeUninit`, but I don't, it's just that I'd have to use `Option<T>` in places where T doesn't implement `Default`. So, using `MaybeUninit` is only added as an optimization but it introduces a risk of causing UB if handled improperly. Although I think that such bugs would surface promptly

Reply:
I don't really see how that can be viewed as an optimisation. If it is, I'd like to see the performance impact before signing off.